internal/runtime/catalog/catalog.go

```diff
@@ -0,0 +1,333 @@
+	gvhToHookDescriptor map[GroupVersionHook]hookDescriptor
+
+	// gvToOpenAPIDefinitions allows mapping a GroupVersion to a func
+	// which returns OpenAPI definitions for all request and response types with that
+	// GroupVersion.
+	gvToOpenAPIDefinitions map[schema.GroupVersion]OpenAPIDefinitionsGetter
+
+	// catalogName is the name of this catalog. It is set based on the stack of the New caller.
+	// This is useful for error reporting to indicate the origin of the Catalog.
+	catalogName string
+}
+
+// Hook is a marker interface for a hook.
+// Hooks should be defined as a: func(*RequestType, *ResponseType).
+type Hook interface{}
+
+// hookDescriptor is an internal data structure which holds
+// all information about a Hook.
+type hookDescriptor struct {
+	metadata *HookMeta
+
+	// request gvk for the Hook.
+	request schema.GroupVersionKind
+	// response gvk for the Hook.
+	response schema.GroupVersionKind
+}
+
+// HookMeta holds metadata for a Hook, which is used to generate
+// the OpenAPI definition for a Hook.
+type HookMeta struct {
+	// Summary of the hook.
+	Summary string
+
+	// Description of the hook.
+	Description string
+
+	// Tags of the hook.
+	Tags []string
+
+	// Deprecated signals if the hook is deprecated.
+	Deprecated bool
+
+	// Singleton signals if the hook can only be implemented once on a
+	// Runtime Extension, e.g. like the DiscoveryHook.
+	Singleton bool
+}
+
+// OpenAPIDefinitionsGetter defines a func which returns OpenAPI definitions for all
+// request and response types of a GroupVersion.
+// NOTE: The OpenAPIDefinitionsGetter funcs in the API packages are generated by openapi-gen.
+type OpenAPIDefinitionsGetter func(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition
+
+// New creates a new Catalog.
+func New() *Catalog {
+	return &Catalog{
+		scheme:                 runtime.NewScheme(),
+		gvhToType:              map[GroupVersionHook]reflect.Type{},
+		typeToGVH:              map[reflect.Type]GroupVersionHook{},
+		gvhToHookDescriptor:    map[GroupVersionHook]hookDescriptor{},
+		gvToOpenAPIDefinitions: map[schema.GroupVersion]OpenAPIDefinitionsGetter{},
+		// Note: We have to ignore the current file so that GetNameFromCallsite retrieves the name of the caller (the parent).
+		catalogName: naming.GetNameFromCallsite("sigs.k8s.io/cluster-api/internal/runtime/catalog/catalog.go"),
+	}
+}
+
+// AddHook adds a HookFunc and its request and response types with the gv GroupVersion.
+// The passed in hookFunc must have the following type: func(*RequestType,*ResponseType)
+// The name of the func becomes the "hook" in GroupVersionHook.
+// GroupVersion must not have empty fields.
+func (c *Catalog) AddHook(gv schema.GroupVersion, hookFunc Hook, hookMeta *HookMeta) {
+	// Validate gv.Group and gv.Version are not empty.
+	if gv.Group == "" {
+		panic("Group must not be empty")
+	}
+	if gv.Version == "" {
+		panic("Version must not be empty")
```

Comment:
Let's keep it as is now, but just fyi. Those are not errors. Those are strings which are handed over to panics. I used upper case because the upstream scheme package was using uppercase consistently.

The reason why errors are lower case is because they are concatenated like this `<error-message>: <error-message>: <error-message>`. Uppercase would look awkward there.

This does not apply to panics.